tower-http/src/catch_panic.rs

```diff
@@ -0,0 +1,202 @@
+//! TODO(david): docs
+
+use bytes::Bytes;
+use futures_core::ready;
+use futures_util::future::{CatchUnwind, FutureExt};
+use http::{Request, Response, StatusCode};
+use http_body::{combinators::UnsyncBoxBody, Body, Full};
+use pin_project_lite::pin_project;
+use std::{
+    any::Any,
+    borrow::Cow,
+    future::Future,
+    panic::AssertUnwindSafe,
+    pin::Pin,
+    task::{Context, Poll},
+};
+use tower_layer::Layer;
+use tower_service::Service;
+
+/// Layer that applies the [`CatchPanic`] middleware that catches panics and converts them into
+/// `500 Internal Server` responses.
+///
+/// See the [module docs](self) for an example.
+#[derive(Debug, Clone, Copy, Default)]
+#[non_exhaustive]
+pub struct CatchPanicLayer;
+
+impl CatchPanicLayer {
+    /// Create a new `CatchPanicLayer`.
+    pub fn new() -> Self {
+        CatchPanicLayer {}
+    }
+}
+
+impl<S> Layer<S> for CatchPanicLayer {
+    type Service = CatchPanic<S>;
+
+    fn layer(&self, inner: S) -> Self::Service {
+        CatchPanic::new(inner)
+    }
+}
+
+/// Middleware that catches panics and converts them into `500 Internal Server` responses.
+///
+/// See the [module docs](self) for an example.
+#[derive(Debug, Clone, Copy)]
+pub struct CatchPanic<S> {
+    inner: S,
+}
+
+impl<S> CatchPanic<S> {
+    /// Create a new `CatchPanic`.
+    pub fn new(inner: S) -> Self {
+        Self { inner }
+    }
+
+    define_inner_service_accessors!();
+
+    /// Returns a new [`Layer`] that wraps services with a `CatchPanic` middleware.
+    ///
+    /// [`Layer`]: tower_layer::Layer
+    pub fn layer() -> CatchPanicLayer {
+        CatchPanicLayer::new()
+    }
+}
+
+impl<S, ReqBody, ResBody> Service<Request<ReqBody>> for CatchPanic<S>
+where
+    S: Service<Request<ReqBody>, Response = Response<ResBody>>,
+    ResBody: Body<Data = Bytes> + Send + 'static,
+{
+    type Response = Response<UnsyncBoxBody<Bytes, ResBody::Error>>;
+    type Error = S::Error;
+    type Future = ResponseFuture<S::Future>;
+
+    #[inline]
+    fn poll_ready(&mut self, cx: &mut Context<'_>) -> Poll<Result<(), Self::Error>> {
+        self.inner.poll_ready(cx)
+    }
+
+    fn call(&mut self, req: Request<ReqBody>) -> Self::Future {
+        match std::panic::catch_unwind(AssertUnwindSafe(|| self.inner.call(req))) {
+            Ok(future) => ResponseFuture {
+                kind: Kind::Future {
+                    future: AssertUnwindSafe(future).catch_unwind(),
```

Comment:
Requiring the service and future to be `UnwindSafe` doesn't work. It doesn't compile when you actually try and use it.

I'm honestly not very familiar with `AssertUnwindSafe` but I suppose this is fine since there is no `unsafe` code 🤔 

@jplatte maybe you know?

Reply:
I had the same experience, but it seems this approach is correct:

* see https://doc.rust-lang.org/stable/src/core/panic/unwind_safe.rs.html#170
* [`catch_unwind` from `futures` basically does the same thing](https://docs.rs/futures-util/0.3.19/src/futures_util/future/future/catch_unwind.rs.html#36)

Reply:
Yeah `AssertUnwindSafe` isn't exactly intuitive. There's even a Pre-RFC about deprecating it where you can read more about what it does, how it's useful and how it's not so useful in other cases: https://internals.rust-lang.org/t/pre-rfc-deprecating-unwindsafe/15974

Reply:
Even thought there might not be soundness issues it looks to me like it could cause potential state corruption in certain cases. I am certainly not familiar enough with catch_unwind to see if this would have the kind of problem but this feels a bit scary